### PR TITLE
[Draft/ReferenceOnly] Port move_group node to ROS2

### PR DESCRIPTION
**Pull request by ddengster**
This PR is done in conjunction with a 'coffeebot' gazebo simulation in which I use move_group to plan and execute trajectories. There will be some stuff missing that should be in the port, I'll try my best to highlight them. Feel free to siphon changes from this PR into the main branch as I will be trying to get the pipeline that decides how to get the poses into the srdf file working before I start merging changes from this PR.

There are other dependencies that have features that needed to be ported to ROS2. These are in the PRs:
- [joint_trajectory_controller](https://github.com/ros-controls/ros2_controllers/pull/36), for updating the state of the robot.
- [gazebo_ros_control](https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1083), for use with the gazebo simulator

### Description

Port move_group node to ROS2. Submodules like capabilities have also been ported. Submodules regarding sensors have not been ported

Design problems regarding nodes:

- Parameters and parameter file handling. Move_group used to have a number of classes that share parameters from a central ROS1 server. I've put them into the move_group node and that node flows throughout different classes in the codebase. The side effect is that you will get a number of deadlocks because that node is also used for subscribing and listening to the same topic. To fix this you'll have create or supply a different node. (see TrajectoryExecutionManager::initialize)

- There's a good number of functions that uses rclcpp::spin_until_future_complete. It unfortunately gives me exceptions ('Node already added to executor') when I already have an existing executor to manage the execution of these nodes for me (see move_group.cpp main function). I recommend not using spin_until_future_complete.

To summarize, having separate nodes for each class has its pros and cons. This PR's style is having 1 node to contain all parameters and making new nodes and supplying them parameters as needed. 

Modifications to launch and config files generated by moveit_setup_assistant:

- There's a ton of xml launch files that I ported to a single python launch file just so things could be easier. In my experience the ROS2 xml launch repo isnt too stable for my liking, and xml is a data format ill-suited to the task that launching programs require. I recommend porting to python launch files.

- config files are prepended with the following as listed in the ROS2 launch wiki:
```
<node name, usually 'move_group'>:
  ros__parameters:
```

### CODE REVIEW

File: moveit_planners/ompl/ompl_interface/src/ompl_interface.cpp [hunk 1]
```diff
@@ -177,32 +182,30 @@ void ompl_interface::OMPLInterface::loadPlannerConfigurations()
     {
       if (node_->has_parameter(group_name_param + "." + k))
       {
-        std::string value;
-        if (node_->get_parameter(group_name_param + "." + k, value))
+        rclcpp::Parameter param = node_->get_parameter(group_name_param + "." + k);
+        if (param.get_type() == rclcpp::ParameterType::PARAMETER_STRING)
```
Comment (ddengster) [inline moveit_planners/ompl/ompl_interface/src/ompl_interface.cpp:186]: same fix mentioned in issue #184

File: moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp [hunk 2]
```diff
@@ -135,4 +134,7 @@ class OMPLPlannerManager : public planning_interface::PlannerManager
 
 }  // namespace ompl_interface
 
-CLASS_LOADER_REGISTER_CLASS(ompl_interface::OMPLPlannerManager, planning_interface::PlannerManager)
+#include <pluginlib/class_list_macros.hpp>
+
+PLUGINLIB_EXPORT_CLASS(
+  ompl_interface::OMPLPlannerManager, planning_interface::PlannerManager)
```
Comment (ddengster) [inline moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp:140]: made OMPLPlannerManager switch to plugin system

File: moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp [hunk 3]
```diff
@@ -779,92 +757,151 @@ class MoveGroupInterface::MoveGroupInterfaceImpl
...
-    {
-      ROS_INFO_STREAM_NAMED("move_group_interface", move_action_client_->getState().toString()
-                                                        << ": " << move_action_client_->getState().getText());
-      return MoveItErrorCode(move_action_client_->getResult()->error_code);
-    }
+    auto res = result_future.get();
+    return res.result->error_code;
   }
 
   MoveItErrorCode execute(const Plan& plan, bool wait)
```
Comment (ddengster) [inline moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp:786]: function of major interest

File: moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp [hunk 4]
```diff
@@ -695,82 +630,125 @@ class MoveGroupInterface::MoveGroupInterfaceImpl
...
+    response = res.get();
+    if (response->error_code.val != moveit_msgs::msg::MoveItErrorCodes::SUCCESS)
+    {
+      RCLCPP_ERROR(LOGGER, "Grasp planning failed. Unable to pick.");
+      return MoveItErrorCode(moveit_msgs::msg::MoveItErrorCodes::FAILURE);
+    }
+    return pick(constructPickupGoal(object.id, std::move(response->grasps), plan_only));
   }
 
   MoveItErrorCode plan(Plan& plan)
```
Comment (ddengster) [inline moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp:665]: function of major interest

File: moveit_ros/move_group/src/move_group.cpp [hunk 5]
```diff
@@ -169,29 +178,73 @@ class MoveGroupExe
...
+  opt.allow_undeclared_parameters(true);
+  opt.automatically_declare_parameters_from_overrides(true);
+  rclcpp::Node::SharedPtr nh = rclcpp::Node::make_shared("move_group", opt);
+
+  //fetch a bunch of parameters
+  {
+    std::string robot_desc_param = "robot_description";
+    std::string str = getParameterFromRemoteNode<std::string>(nh, "robot_state_publisher", robot_desc_param);
+    nh->declare_parameter(robot_desc_param);
+    nh->set_parameter(rclcpp::Parameter(robot_desc_param, str));
```
[the remark this answers is not in the document]
Comment (ddengster) [reply moveit_ros/move_group/src/move_group.cpp:226]: to clarify 'my other codebases' meaning the codebases that use MoveGroupInterface or PlanningSceneInterface

**Comment by henningkayser** [comment]
@ddengster first of all, great work and thanks a lot for your effort! We were actually about to spin-up development again to make progress towards the full release, so I'm really appreciating your contributions. 

> This PR is done in conjunction with a 'coffeebot' gazebo simulation in which I use move_group to plan and execute trajectories. There will be some stuff missing that should be in the port, I'll try my best to highlight them. Feel free to siphon changes from this PR into the main branch as I will be trying to get the pipeline that decides how to get the poses into the srdf file working before I start merging changes from this PR.

In general, we do fresh package/library migrations piece by piece so full pipeline functionality is not required out of the box.

> 
> There are other dependencies that have features that needed to be ported to ROS2. These are in the PRs:
> 
> * [joint_trajectory_controller](https://github.com/ros-controls/ros2_controllers/pull/36), for updating the state of the robot.
> * [gazebo_ros_control](https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1083), for use with the gazebo simulator

Nice!

> 
> ### Description
> Port move_group node to ROS2. Submodules like capabilities have also been ported. Submodules regarding sensors have not been ported
> 
> Design problems regarding nodes:
> 
> * Parameters and parameter file handling. Move_group used to have a number of classes that share parameters from a central ROS1 server. I've put them into the move_group node and that node flows throughout different classes in the codebase. The side effect is that you will get a number of deadlocks because that node is also used for subscribing and listening to the same topic. To fix this you'll have create or supply a different node. (see TrajectoryExecutionManager::initialize)

Yes, with MoveGroup we're now at a point where we have to make some non-trivial design decisions. I see two basic approaches: 1. Implement a "global" parameter server that resembles the ROS1 pattern; 2. Refactor parameter propagation between classes and enable parameter syncing between nodes that require it. We're about to investigate these and related changes the next weeks.

> * There's a good number of functions that uses rclcpp::spin_until_future_complete. It unfortunately gives me exceptions ('Node already added to executor') when I already have an existing executor to manage the execution of these nodes for me (see move_group.cpp main function). I recommend not using spin_until_future_complete.

I fully agree, this is a pain point for which we still need to discuss some best practices.

> 
> To summarize, having separate nodes for each class has its pros and cons. This PR's style is having 1 node to contain all parameters and making new nodes and supplying them parameters as needed.

Absolutely the right approach for now. Our goal is to first have a full straight port which can be synced with MoveIt 1 and only then we start making further ROS2 related refactorings.
 
> 
> Modifications to launch and config files generated by moveit_setup_assistant:
> 
> * There's a ton of xml launch files that I ported to a single python launch file just so things could be easier. In my experience the ROS2 xml launch repo isnt too stable for my liking, and xml is a data format ill-suited to the task that launching programs require. I recommend porting to python launch files.
> * config files are prepended with the following as listed in the ROS2 launch wiki:
> 
> ```
> <node name, usually 'move_group'>:
>   ros__parameters:
> ```

Using python launch files is the way to go IMO. That means that we might have to redesign some parts of the MSA which should also include considerations about a potential multi-node setup or how we want to design parameter lookup.

Right now I'm working on a development roadmap and scheduling weekly sync meetings to get the project kicked-off. Since your contributions appear like you are working on this more long-term, are you interested in joining the meetings to get introduced and collaborate on some of the points discussed here?

**Comment by davetcoleman** [comment]
Great effort @ddengster, thanks for sharing this and we look forward to getting it merged!

**Comment by ddengster** [comment]
> Since your contributions appear like you are working on this more long-term, are you interested in joining the meetings to get introduced and collaborate on some of the points discussed here?

Yea I'd like to know who's working on what, and to give you a heads-up as to what I'm working on so as to reduce conflicts. Thing is I'm on GMT+8 Singapore time and might not be able to attend US timings. Do you have a google doc for the meetings so I can at least write down what I'm working on atm or read what's happening?

**Comment by henningkayser** [comment]
> > Since your contributions appear like you are working on this more long-term, are you interested in joining the meetings to get introduced and collaborate on some of the points discussed here?
> 
> Yea I'd like to know who's working on what, and to give you a heads-up as to what I'm working on so as to reduce conflicts. Thing is I'm on GMT+8 Singapore time and might not be able to attend US timings. Do you have a google doc for the meetings so I can at least write down what I'm working on atm or read what's happening?

We don't have the meeting schedule fixed yet, but it will share plans and results publicly here on Github. I'm at GMT+2 myself, but maybe we can find a time that works for all. Could you send me a mail so that we can sync up?

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 6]
```diff
@@ -123,6 +129,7 @@ class ActionBasedControllerHandle : public ActionBasedControllerHandleBase
     {
       RCLCPP_INFO_STREAM(LOGGER, "Cancelling execution for " << name_);
       auto cancel_result_future = controller_action_client_->async_cancel_goal(current_goal_);
+      //@todo: replace rclcpp::spin_until_future_complete
       if (rclcpp::spin_until_future_complete(node_, cancel_result_future) !=
           rclcpp::executor::FutureReturnCode::SUCCESS)
```
[the remark this answers is not in the document]
Comment (henningkayser) [reply moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h:134]: So, for `std::shared_future` results from async calls we should just use `wait()` `wait_for()` or `wait_until()` since the executor is already doing the spinning.

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 7]
```diff
@@ -137,9 +144,12 @@ class ActionBasedControllerHandle : public ActionBasedControllerHandleBase
   bool waitForExecution(const rclcpp::Duration& timeout = rclcpp::Duration(0)) override
   {
     auto result_future = controller_action_client_->async_get_result(current_goal_);
-    if (controller_action_client_ && !done_)
-      return (rclcpp::spin_until_future_complete(node_, result_future, timeout.to_chrono<std::chrono::seconds>()) ==
-              rclcpp::executor::FutureReturnCode::SUCCESS);
+
+    rclcpp::Time start_time = node_->now();
+    auto end_time = start_time + timeout;
```
Comment (henningkayser) [inline moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h:149]: Same here. Also, the default `timeout = rclcpp::Duration(0)` should indicate an infinite timeout

File: moveit_plugins/moveit_simple_controller_manager/src/follow_joint_trajectory_controller_handle.cpp [hunk 8]
```diff
@@ -54,34 +54,42 @@ bool FollowJointTrajectoryControllerHandle::sendTrajectory(const moveit_msgs::ms
   }
 
   if (done_)
-    RCLCPP_DEBUG_STREAM(LOGGER, "sending trajectory to " << name_);
+    RCLCPP_INFO_STREAM(LOGGER, "sending trajectory to " << name_);
```
Comment (henningkayser) [inline moveit_plugins/moveit_simple_controller_manager/src/follow_joint_trajectory_controller_handle.cpp:57]: Why change the log level?

File: moveit_plugins/moveit_simple_controller_manager/src/follow_joint_trajectory_controller_handle.cpp [hunk 9]
```diff
@@ -54,34 +54,42 @@ bool FollowJointTrajectoryControllerHandle::sendTrajectory(const moveit_msgs::ms
...
+    else
+      RCLCPP_INFO(LOGGER, "Goal request accepted!");
   };
   // Result callback
   send_goal_options.result_callback =
       std::bind(&FollowJointTrajectoryControllerHandle::controllerDoneCallback, this, _1);
   // Send goal
   auto current_goal_future = controller_action_client_->async_send_goal(goal, send_goal_options);
-  if (rclcpp::spin_until_future_complete(node_, current_goal_future) != rclcpp::executor::FutureReturnCode::SUCCESS)
+  std::future_status status = current_goal_future.wait_for(std::chrono::seconds(60));
```
Comment (henningkayser) [inline moveit_plugins/moveit_simple_controller_manager/src/follow_joint_trajectory_controller_handle.cpp:79]: Shouldn't this timeout be the expected execution time + some margin?

File: moveit_ros/manipulation/move_group_pick_place_capability/CMakeLists.txt [hunk 10]
```diff
@@ -7,11 +7,11 @@ add_library(${MOVEIT_LIB_NAME}
   )
 set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES VERSION "${${PROJECT_NAME}_VERSION}")
 
-target_link_libraries(${MOVEIT_LIB_NAME} moveit_pick_place_planner ${catkin_LIBRARIES} ${Boost_LIBRARIES})
+target_link_libraries(${MOVEIT_LIB_NAME} moveit_pick_place_planner ${LIBS} ${Boost_LIBRARIES})
```
Comment (henningkayser) [inline moveit_ros/manipulation/move_group_pick_place_capability/CMakeLists.txt:10]: Is Boost still required?

File: moveit_ros/move_group/CMakeLists.txt [hunk 11]
```diff
@@ -9,45 +25,33 @@ if(NOT CMAKE_CONFIGURATION_TYPES AND NOT CMAKE_BUILD_TYPE)
...
-set_target_properties(moveit_move_group_capabilities_base PROPERTIES VERSION "${${PROJECT_NAME}_VERSION}")
-add_dependencies(moveit_move_group_capabilities_base ${catkin_EXPORTED_TARGETS}) # wait until all *_msgs packages are finished being built
+
+add_library(moveit_move_group_capabilities_base SHARED src/move_group_context.cpp
+  src/move_group_capability.cpp)
+set_target_properties(moveit_move_group_capabilities_base
+  PROPERTIES
+  VERSION "${${PROJECT_NAME}_VERSION}")
+
+# Causes the visibility macros to use dllexport rather than dllimport,
```
Comment (henningkayser) [inline moveit_ros/move_group/CMakeLists.txt:35]: I assume you're working on Windows? ;)

File: moveit_ros/move_group/CMakeLists.txt [hunk 12]
```diff
@@ -58,36 +62,52 @@ add_library(moveit_move_group_default_capabilities
...
+  RUNTIME
+  DESTINATION lib/${PROJECT_NAME})
 
-install(TARGETS moveit_move_group_capabilities_base moveit_move_group_default_capabilities
-  ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
-  LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
-  RUNTIME DESTINATION ${CATKIN_GLOBAL_BIN_DESTINATION})
+install(TARGETS moveit_move_group_capabilities_base
+  moveit_move_group_default_capabilities
+  ARCHIVE
```
Comment (henningkayser) [inline moveit_ros/move_group/CMakeLists.txt:82]: Just a nit, but I think it reads a little better with the old format

File: moveit_ros/move_group/include/moveit/move_group/capability_names.h [hunk 13]
```diff
@@ -49,7 +49,7 @@ static const std::string GET_PLANNER_PARAMS_SERVICE_NAME =
     "get_planner_params";  // service name to retrieve planner parameters
 static const std::string SET_PLANNER_PARAMS_SERVICE_NAME =
     "set_planner_params";                                 // service name to set planner parameters
-static const std::string MOVE_ACTION = "move_group";      // name of 'move' action
+static const std::string MOVE_ACTION = "move_action";      // name of 'move' action
```
[the remark this answers is not in the document]
Comment (henningkayser) [reply moveit_ros/move_group/include/moveit/move_group/capability_names.h:52]: Interesting, conflict with what exactly? In general I don't see anything wrong with renaming this, though.

File: moveit_ros/move_group/include/moveit/move_group/move_group_capability.h [hunk 14]
```diff
@@ -57,14 +57,16 @@ MOVEIT_CLASS_FORWARD(MoveGroupCapability)
...
+  MoveGroupCapability(const std::string& capability_name) : capability_name_(capability_name)
   {
   }
-
+  
   virtual ~MoveGroupCapability()
   {
   }
 
+  void setNode(const rclcpp::Node::SharedPtr& node);
```
Comment (henningkayser) [inline moveit_ros/move_group/include/moveit/move_group/move_group_capability.h:68]: This and `setContext()` are only called once before `initialize()`, right? I would pass node and context as parameters of `initialize()` and remove the setters (similar to how the planning adapters are initialized now).

File: moveit_ros/move_group/src/default_capabilities/apply_planning_scene_service_capability.cpp [hunk 15]
```diff
@@ -39,29 +39,41 @@
 
 namespace move_group
 {
+
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_move_group_default_capabilities.apply_planning_scene_service_capability");
```
Comment (henningkayser) [inline moveit_ros/move_group/src/default_capabilities/apply_planning_scene_service_capability.cpp:43]: I see, it's really super long. I would probably just remove the "default_capabilities" in the middle, actually.

**Comment by ddengster** [comment]
> Could you send me a mail so that we can sync up?

I've sent the email

**Comment by ddengster** [comment]
closing since move_group is merged
